File: scripts/build_crosswalk.py
"""
Build Anthropic API Task → O*NET Crosswalk
==========================================

This script creates a crosswalk linking Anthropic's Claude API task classifications
to O*NET occupational task codes and occupation codes (SOC).

Data Sources:
- Anthropic Economic Index API data (November 2025): HuggingFace
- O*NET Database v29.1: https://www.onetcenter.org/database.html

Methodology:
1. Extract unique task descriptions from Anthropic API data
2. Normalize text (lowercase, remove punctuation, normalize whitespace)
3. Exact match to O*NET Task Statements
4. Fuzzy match remaining tasks (threshold >= 85) using rapidfuzz
5. Enrich with O*NET occupation data, Job Zones, and education requirements

Output:
- master_task_crosswalk.csv: Main crosswalk file
- unmatched_tasks.csv: Tasks that could not be matched

Author: Ilan Strauss / AI Disclosures Project
Date: January 2026
"""

import pandas as pd
from rapidfuzz import fuzz, process
import re
import os

# Paths
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
RAW_DIR = os.path.join(BASE_DIR, 'raw')
PROCESSED_DIR = os.path.join(BASE_DIR, 'processed')
ANTHROPIC_DATA = os.path.join(BASE_DIR, '..', 'release_2026_01_15', 'data', 'intermediate',
                               'aei_raw_1p_api_2025-11-13_to_2025-11-20.csv')
ONET_DIR = os.path.join(RAW_DIR, 'db_29_1_text')

# Fuzzy match threshold
FUZZY_THRESHOLD = 85


def normalize_text(text):
    """Normalize text for matching: lowercase, remove punctuation, normalize whitespace."""
    text = str(text).lower().strip()
    text = re.sub(r'[^\w\s]', '', text)
    text = re.sub(r'\s+', ' ', text)
    return text


def load_data():
    """Load Anthropic and O*NET data."""
    print("Loading data...")

    # Anthropic API data
    anthropic = pd.read_csv(ANTHROPIC_DATA)

    # O*NET data
    onet_tasks = pd.read_csv(os.path.join(ONET_DIR, 'Task Statements.txt'), sep='\t')
    onet_occs = pd.read_csv(os.path.join(ONET_DIR, 'Occupation Data.txt'), sep='\t')
    job_zones = pd.read_csv(os.path.join(ONET_DIR, 'Job Zones.txt'), sep='\t')
    education = pd.read_csv(os.path.join(ONET_DIR, 'Education, Training, and Experience.txt'), sep='\t')

    return anthropic, onet_tasks, onet_occs, job_zones, education


def extract_anthropic_tasks(anthropic):
    """Extract task descriptions and counts from Anthropic data."""
    print("Extracting Anthropic tasks...")

    task_data = anthropic[
        (anthropic['facet'] == 'onet_task') &
        (anthropic['variable'] == 'onet_task_count')
    ][['cluster_name', 'value']].copy()

    task_data.columns = ['anthropic_task', 'api_count']

    # Exclude placeholder categories
    non_tasks = ['not_classified', 'none']
    task_data = task_data[~task_data['anthropic_task'].str.lower().isin(non_tasks)]

    task_data['task_norm'] = task_data['anthropic_task'].apply(normalize_text)

    print(f"  Found {len(task_data)} unique tasks")
    return task_data


def exact_match(task_data, onet_tasks):
    """Perform exact matching on normalized text."""
    print("Performing exact matching...")

    onet_tasks['task_norm'] = onet_tasks['Task'].apply(normalize_text)
    onet_lookup = onet_tasks.groupby('task_norm').first().reset_index()[
        ['task_norm', 'O*NET-SOC Code', 'Task ID', 'Task', 'Task Type']
    ]

    merged = task_data.merge(onet_lookup, on='task_norm', how='left')
    matched = merged[merged['O*NET-SOC Code'].notna()].copy()
    matched['match_method'] = 'exact'
    matched['match_score'] = 100.0

    unmatched = merged[merged['O*NET-SOC Code'].isna()][
        ['anthropic_task', 'api_count', 'task_norm']
    ]

    print(f"  Exact matches: {len(matched)}")
    print(f"  Remaining for fuzzy matching: {len(unmatched)}")

    return matched, unmatched, onet_tasks


def fuzzy_match(unmatched, onet_tasks, threshold=FUZZY_THRESHOLD):
    """Perform fuzzy matching on remaining tasks."""
    print(f"Performing fuzzy matching (threshold >= {threshold})...")

    onet_choices = list(onet_tasks['task_norm'].unique())
    onet_norm_lookup = onet_tasks.groupby('task_norm').first().reset_index()

    fuzzy_rows = []
    for _, row in unmatched.iterrows():
        result = process.extractOne(row['task_norm'], onet_choices, scorer=fuzz.ratio)
        if result and result[1] >= threshold:
            match_text, score, _ = result
            onet_match = onet_norm_lookup[onet_norm_lookup['task_norm'] == match_text]
            if len(onet_match) > 0:
                onet_row = onet_match.iloc[0]
                fuzzy_rows.append({
                    'anthropic_task': row['anthropic_task'],
                    'api_count': row['api_count'],
                    'task_norm': row['task_norm'],
                    'O*NET-SOC Code': onet_row['O*NET-SOC Code'],
                    'Task ID': onet_row['Task ID'],
                    'Task': onet_row['Task'],
                    'Task Type': onet_row['Task Type'],
                    'match_method': 'fuzzy',
                    'match_score': score
                })

    fuzzy_matched = pd.DataFrame(fuzzy_rows)
    print(f"  Fuzzy matches: {len(fuzzy_matched)}")

    return fuzzy_matched


def enrich_with_onet(matched, onet_occs, job_zones, education):
    """Add O*NET occupation data, job zones, and education."""
    print("Enriching with O*NET attributes...")

    # Add occupation titles and descriptions
    matched = matched.merge(
        onet_occs[['O*NET-SOC Code', 'Title', 'Description']],
        on='O*NET-SOC Code', how='left'
    )

    # Add job zones
    jz_clean = job_zones[['O*NET-SOC Code', 'Job Zone']].drop_duplicates()
    matched = matched.merge(jz_clean, on='O*NET-SOC Code', how='left')

    # Add typical education
    edu_level = education[education['Element Name'] == 'Required Level of Education'][
        ['O*NET-SOC Code', 'Category', 'Data Value']
    ]
    edu_pivot = edu_level.pivot_table(
        index='O*NET-SOC Code', columns='Category',
        values='Data Value', aggfunc='first'
    ).reset_index()

    edu_categories = [c for c in edu_pivot.columns if c != 'O*NET-SOC Code']
    if edu_categories:
        edu_pivot['typical_education'] = edu_pivot[edu_categories].idxmax(axis=1)
        edu_pivot['typical_education_pct'] = edu_pivot[edu_categories].max(axis=1)
        matched = matched.merge(
            edu_pivot[['O*NET-SOC Code', 'typical_education', 'typical_education_pct']],
            on='O*NET-SOC Code', how='left'
        )

    return matched


def save_outputs(matched, task_data, output_dir):
    """Save final crosswalk and unmatched tasks."""
    print("Saving outputs...")

    # Rename columns for final output
    final = matched[[
        'anthropic_task', 'api_count',
        'O*NET-SOC Code', 'Task ID', 'Task', 'Task Type',
        'Title', 'Description',
        'match_method', 'match_score',
        'Job Zone', 'typical_education', 'typical_education_pct'
    ]].copy()

    final.columns = [
        'anthropic_task_description', 'api_usage_count',
        'onet_soc_code', 'onet_task_id', 'onet_task_description', 'onet_task_type',
        'onet_occupation_title', 'onet_occupation_description',
        'match_method', 'match_score',
        'job_zone', 'typical_education', 'typical_education_pct'
    ]

    final = final.sort_values('api_usage_count', ascending=False)
    final.to_csv(os.path.join(output_dir, 'master_task_crosswalk.csv'), index=False)

    # Save unmatched
    matched_tasks = set(matched['anthropic_task'])
    unmatched_final = task_data[~task_data['anthropic_task'].isin(matched_tasks)]
    unmatched_final = unmatched_final.sort_values('api_count', ascending=False)
    unmatched_final[['anthropic_task', 'api_count']].to_csv(
        os.path.join(output_dir, 'unmatched_tasks.csv'), index=False
    )

    return final, unmatched_final


def print_summary(final, task_data, unmatched_final):
    """Print summary statistics."""
    print("\n" + "="*60)
    print("CROSSWALK SUMMARY")
    print("="*60)
    print(f"Total Anthropic tasks: {len(task_data)}")
    print(f"Matched tasks: {len(final)}")
    print(f"Match rate: {len(final) / len(task_data) * 100:.1f}%")
    print(f"\nAPI usage coverage: {final['api_usage_count'].sum() / task_data['api_count'].sum() * 100:.1f}%")
    print(f"\nMatch method breakdown:")
    print(final['match_method'].value_counts().to_string())
    print(f"\nUnique SOC codes: {final['onet_soc_code'].nunique()}")
    print(f"Unmatched tasks: {len(unmatched_final)}")


def main():
    """Main execution."""
    print("="*60)
    print("Building Anthropic → O*NET Task Crosswalk")
    print("="*60 + "\n")

    # Load data
    anthropic, onet_tasks, onet_occs, job_zones, education = load_data()

    # Extract Anthropic tasks
    task_data = extract_anthropic_tasks(anthropic)

    # Exact matching
    exact_matched, unmatched, onet_tasks = exact_match(task_data, onet_tasks)

    # Fuzzy matching
    fuzzy_matched = fuzzy_match(unmatched, onet_tasks)

    # Combine matches
    all_matched = pd.concat([exact_matched, fuzzy_matched], ignore_index=True)

    # Enrich with O*NET data
    enriched = enrich_with_onet(all_matched, onet_occs, job_zones, education)

    # Save outputs
    final, unmatched_final = save_outputs(enriched, task_data, PROCESSED_DIR)

    # Print summary
    print_summary(final, task_data, unmatched_final)

    print("\n" + "="*60)
    print("Done!")
    print("="*60)


if __name__ == '__main__':
    main()

File: R/build_crosswalk.R
# =============================================================================
# Build Anthropic API Task → O*NET Crosswalk
# =============================================================================
#
# Links Anthropic Claude API task descriptions to O*NET occupational codes via:
# 1. Text normalization (lowercase, remove punctuation)
# 2. Exact string matching
# 3. Fuzzy matching (Levenshtein, threshold >= 85)
# 4. Enrichment with O*NET occupation attributes
#
# Author: Ilan Strauss | AI Disclosures Project
# Date: January 2026
# =============================================================================

library(tidyverse)
library(stringdist)  # For fuzzy matching

# --- CONFIGURATION ---
BASE_DIR <- dirname(dirname(rstudioapi::getSourceEditorContext()$path))
# If not in RStudio, use:
# BASE_DIR <- here::here()

RAW_DIR <- file.path(BASE_DIR, "raw")
PROCESSED_DIR <- file.path(BASE_DIR, "processed")
ANTHROPIC_DATA <- file.path(BASE_DIR, "..", "release_2026_01_15", "data", "intermediate",
                            "aei_raw_1p_api_2025-11-13_to_2025-11-20.csv")
ONET_DIR <- file.path(RAW_DIR, "db_29_1_text")

FUZZY_THRESHOLD <- 85  # Minimum similarity score (0-100)


# =============================================================================
# TEXT NORMALIZATION
# =============================================================================
# Lowercase, remove punctuation, collapse whitespace

normalize_text <- function(text) {
  text %>%
    tolower() %>%
    str_trim() %>%
    str_replace_all("[^[:alnum:][:space:]]", "") %>%
    str_replace_all("\\s+", " ")
}


# =============================================================================
# LOAD DATA
# =============================================================================

load_data <- function() {
  anthropic <- read_csv(ANTHROPIC_DATA, show_col_types = FALSE)

  onet_tasks <- read_tsv(file.path(ONET_DIR, "Task Statements.txt"), show_col_types = FALSE)
  onet_occs <- read_tsv(file.path(ONET_DIR, "Occupation Data.txt"), show_col_types = FALSE)
  job_zones <- read_tsv(file.path(ONET_DIR, "Job Zones.txt"), show_col_types = FALSE)
  education <- read_tsv(file.path(ONET_DIR, "Education, Training, and Experience.txt"), show_col_types = FALSE)

  list(
    anthropic = anthropic,
    onet_tasks = onet_tasks,
    onet_occs = onet_occs,
    job_zones = job_zones,
    education = education
  )
}


# =============================================================================
# EXTRACT ANTHROPIC TASKS
# =============================================================================
# Filter to O*NET task facet and extract task descriptions with API counts

extract_anthropic_tasks <- function(anthropic) {
  anthropic %>%
    filter(facet == "onet_task", variable == "onet_task_count") %>%
    select(anthropic_task = cluster_name, api_count = value) %>%
    filter(!tolower(anthropic_task) %in% c("not_classified", "none")) %>%
    mutate(task_norm = normalize_text(anthropic_task))
}


# =============================================================================
# EXACT MATCHING
# =============================================================================
# Match normalized Anthropic tasks to O*NET tasks via exact string comparison

exact_match <- function(task_data, onet_tasks) {
  # Normalize O*NET tasks
  onet_tasks <- onet_tasks %>%
    mutate(task_norm = normalize_text(Task))

  # Create lookup (first occurrence per normalized text)
  onet_lookup <- onet_tasks %>%
    group_by(task_norm) %>%
    slice(1) %>%
    ungroup() %>%
    select(task_norm, `O*NET-SOC Code`, `Task ID`, Task, `Task Type`)

  # Join
  merged <- task_data %>%
    left_join(onet_lookup, by = "task_norm")

  # Split matched and unmatched
  matched <- merged %>%
    filter(!is.na(`O*NET-SOC Code`)) %>%
    mutate(match_method = "exact", match_score = 100)

  unmatched <- merged %>%
    filter(is.na(`O*NET-SOC Code`)) %>%
    select(anthropic_task, api_count, task_norm)

  list(matched = matched, unmatched = unmatched, onet_tasks = onet_tasks)
}


# =============================================================================
# FUZZY MATCHING
# =============================================================================
# Match remaining tasks using Levenshtein similarity

fuzzy_match <- function(unmatched, onet_tasks, threshold = FUZZY_THRESHOLD) {
  # Get unique O*NET normalized tasks
  onet_lookup <- onet_tasks %>%
    group_by(task_norm) %>%
    slice(1) %>%
    ungroup()

  onet_choices <- unique(onet_lookup$task_norm)

  # Match each unmatched task
  fuzzy_results <- map_dfr(seq_len(nrow(unmatched)), function(i) {
    query <- unmatched$task_norm[i]

    # Calculate Levenshtein similarity (convert distance to 0-100 score)
    distances <- stringdist(query, onet_choices, method = "lv")
    max_len <- pmax(nchar(query), nchar(onet_choices))
    scores <- 100 * (1 - distances / max_len)

    best_idx <- which.max(scores)
    best_score <- scores[best_idx]

    if (best_score >= threshold) {
      best_match <- onet_choices[best_idx]
      onet_row <- onet_lookup %>% filter(task_norm == best_match) %>% slice(1)

      tibble(
        anthropic_task = unmatched$anthropic_task[i],
        api_count = unmatched$api_count[i],
        task_norm = query,
        `O*NET-SOC Code` = onet_row$`O*NET-SOC Code`,
        `Task ID` = onet_row$`Task ID`,
        Task = onet_row$Task,
        `Task Type` = onet_row$`Task Type`,
        match_method = "fuzzy",
        match_score = best_score
      )
    } else {
      NULL
    }
  })

  fuzzy_results
}


# =============================================================================
# ENRICH WITH O*NET ATTRIBUTES
# =============================================================================
# Add occupation titles, job zones, and typical education

enrich_with_onet <- function(matched, onet_occs, job_zones, education) {
  # Add occupation titles and descriptions
  matched <- matched %>%
    left_join(
      onet_occs %>% select(`O*NET-SOC Code`, Title, Description),
      by = "O*NET-SOC Code"
    )

  # Add job zones (1-5 education/preparation scale)
  jz_clean <- job_zones %>%
    select(`O*NET-SOC Code`, `Job Zone`) %>%
    distinct()

  matched <- matched %>%
    left_join(jz_clean, by = "O*NET-SOC Code")

  # Add typical education level
  edu_level <- education %>%
    filter(`Element Name` == "Required Level of Education") %>%
    select(`O*NET-SOC Code`, Category, `Data Value`)

  if (nrow(edu_level) > 0) {
    edu_pivot <- edu_level %>%
      pivot_wider(names_from = Category, values_from = `Data Value`, values_fn = first)

    edu_cols <- setdiff(names(edu_pivot), "O*NET-SOC Code")
    if (length(edu_cols) > 0) {
      edu_pivot <- edu_pivot %>%
        rowwise() %>%
        mutate(
          typical_education = edu_cols[which.max(c_across(all_of(edu_cols)))],
          typical_education_pct = max(c_across(all_of(edu_cols)), na.rm = TRUE)
        ) %>%
        ungroup() %>%
        select(`O*NET-SOC Code`, typical_education, typical_education_pct)

      matched <- matched %>%
        left_join(edu_pivot, by = "O*NET-SOC Code")
    }
  }

  matched
}


# =============================================================================
# SAVE OUTPUTS
# =============================================================================

save_outputs <- function(matched, task_data, output_dir) {
  # Rename and select final columns
  final <- matched %>%
    select(
      anthropic_task_description = anthropic_task,
      api_usage_count = api_count,
      onet_soc_code = `O*NET-SOC Code`,
      onet_task_id = `Task ID`,
      onet_task_description = Task,
      onet_task_type = `Task Type`,
      onet_occupation_title = Title,
      onet_occupation_description = Description,
      match_method,
      match_score,
      job_zone = `Job Zone`,
      typical_education,
      typical_education_pct
    ) %>%
    arrange(desc(api_usage_count))

  write_csv(final, file.path(output_dir, "master_task_crosswalk.csv"))

  # Save unmatched tasks
  matched_tasks <- unique(matched$anthropic_task)
  unmatched_final <- task_data %>%
    filter(!anthropic_task %in% matched_tasks) %>%
    select(anthropic_task, api_count) %>%
    arrange(desc(api_count))

  write_csv(unmatched_final, file.path(output_dir, "unmatched_tasks.csv"))

  list(final = final, unmatched = unmatched_final)
}


# =============================================================================
# MAIN EXECUTION
# =============================================================================

main <- function() {
  # Load data
  data <- load_data()

  # Extract Anthropic tasks
  task_data <- extract_anthropic_tasks(data$anthropic)

  # Exact matching
  exact_results <- exact_match(task_data, data$onet_tasks)

  # Fuzzy matching (this may take a few minutes)
  fuzzy_matched <- fuzzy_match(exact_results$unmatched, exact_results$onet_tasks)

  # Combine matches
  all_matched <- bind_rows(exact_results$matched, fuzzy_matched)

  # Enrich with O*NET attributes
  enriched <- enrich_with_onet(all_matched, data$onet_occs, data$job_zones, data$education)

  # Save outputs
  save_outputs(enriched, task_data, PROCESSED_DIR)
}

# Run if executed directly
if (sys.nframe() == 0) {
  main()
}
